Enhancement: Allow no-extraneous-dependencies to check internal Imports

In our monorepo I have setup that anything prefixed with out npm-scope is marked as internal, ie. `^@scope-.*/`. This is working really well for us.

We are now at a stage where we are going to need to publish some of our packages to be consumed in other projects/repos. Due to this I now wish to use `no-extraneous-dependencies` to ensure that everyone is rembering to include this internal package references.

I did consider doing this by removing the internal regex, but this then effects the import sortinng (as we keep our internal deps grouped seperatly to our external deps - thanks to the `order` rule)

What I would like to see is to add an option to `no-extraneous-dependencies` to allow it to check internal imports based on a regex (either the internal regex or a newly defined on the rule config) as well as external.

I would be happy to work on a PR to add this option.

I'm having the exact opposite problem. Maybe we can merge our configs?

Right now, when I lint from my project root, I'm getting this error:

> /Users/msc/Code/hub/@arc/infra/cdk/node_modules/@arc/view-app/node_modules/@arc/ui/hooks/useWindowScroll.js
  3:1  error  '@arc/ui' should be listed in the project's dependencies. Run 'npm i -S @arc/ui' to add it  import/no-extraneous-dependencies

`no-extraneous-dependencies` can detect that @arc/cdk must include @arc/view-app in it's deps. But it's throwing an error because @arc/cdk doesn't directly rely on @arc/ui, while @arc/view-app does. 

My relevant config looks like this:

```js
// import

const path = require('path');

const fs = require('fs-extra');
const globby = require('globby');

// vars

const cwd = process.cwd();
const root = path.resolve(__dirname, '../../../');
const lernaFile = path.resolve(cwd, './lerna.json');

const lernaGlobs = fs.existsSync(lernaFile) ? fs.readJsonSync(lernaFile).packages : [];
const pkgPaths = lernaGlobs
  .flatMap((loc) => globby.sync([loc, '!**/node_modules/**'], {onlyDirectories: true, expandDirectories: false}))
  .map((loc) => path.resolve(cwd, loc));

const pkgContainers = fs.readJsonSync(path.resolve(root, 'lerna.json')).packages
  .map((loc) => path.resolve(root, loc.replace(/(\/|^)[^/]+$/, '')));

// export

module.exports = {
  plugins: ['import'],

  settings: {
    'import/resolver': {
      'node': {
        extensions: ['.js', '.jsx'],
        paths: [cwd, ...pkgPaths],
      },
      'eslint-import-resolver-lerna': {
        packages: pkgContainers,
      },
    },
    'import/ignore': [/\.es\.js$/],
  },

  rules: {
    'import/no-extraneous-dependencies': 2
  },
};
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ This change log adheres to standards from [Keep a CHANGELOG](https://keepachange
 - [`no-restricted-paths`]: support arrays for `from` and `target` options ([#2466], thanks [@AdriAt360])
 - [`no-anonymous-default-export`]: add `allowNew` option ([#2505], thanks [@DamienCassou])
 - [`order`]: Add `distinctGroup` option ([#2395], thanks [@hyperupcall])
+- [`no-extraneous-dependencies`]: Add `includeInternal` option ([#2541], thanks [@bdwain])
 
 ### Fixed
 - [`order`]: move nested imports closer to main import entry ([#2396], thanks [@pri1311])
@@ -1008,6 +1009,7 @@ for info on changes for earlier releases.
 
 [`memo-parser`]: ./memo-parser/README.md
 
+[#2541]: https://github.com/import-js/eslint-plugin-import/pull/2541
 [#2531]: https://github.com/import-js/eslint-plugin-import/pull/2531
 [#2511]: https://github.com/import-js/eslint-plugin-import/pull/2511
 [#2506]: https://github.com/import-js/eslint-plugin-import/pull/2506

--- a/docs/rules/no-extraneous-dependencies.md
+++ b/docs/rules/no-extraneous-dependencies.md
@@ -1,7 +1,7 @@
 # import/no-extraneous-dependencies: Forbid the use of extraneous packages
 
 Forbid the import of external modules that are not declared in the `package.json`'s `dependencies`, `devDependencies`, `optionalDependencies`, `peerDependencies`, or `bundledDependencies`.
-The closest parent `package.json` will be used. If no `package.json` is found, the rule will not lint anything. This behavior can be changed with the rule option `packageDir`.
+The closest parent `package.json` will be used. If no `package.json` is found, the rule will not lint anything. This behavior can be changed with the rule option `packageDir`. Normally ignores imports of modules marked internal, but this can be changed with the rule option `includeInternal`.
 
 Modules have to be installed for this rule to work.
 
@@ -30,6 +30,12 @@ You can also use an array of globs instead of literal booleans:
 ```
 
 When using an array of globs, the setting will be set to `true` (no errors reported) if the name of the file being linted matches a single glob in the array, and `false` otherwise.
+
+There is a boolean option called `includeInternal`, which enables the checking of internal modules, which are otherwise ignored by this rule.
+
+```js
+"import/no-extraneous-dependencies": ["error", {"includeInternal": true}]
+```
 
 Also there is one more option called `packageDir`, this option is to specify the path to the folder containing package.json.
 
@@ -99,6 +105,10 @@ var isArray = require('lodash.isarray');
 /* eslint import/no-extraneous-dependencies: ["error", {"bundledDependencies": false}] */
 import foo from '"@generated/foo"';
 var foo = require('"@generated/foo"');
+
+/* eslint import/no-extraneous-dependencies: ["error", {"includeInternal": true}] */
+import foo from './foo';
+var foo = require('./foo');
 ```
 
 

--- a/src/rules/no-extraneous-dependencies.js
+++ b/src/rules/no-extraneous-dependencies.js
@@ -183,7 +183,12 @@ function reportIfMissing(context, deps, depsOptions, node, name) {
     return;
   }
 
-  if (importType(name, context) !== 'external') {
+  const typeOfImport = importType(name, context);
+
+  if (
+    typeOfImport !== 'external'
+    && (typeOfImport !== 'internal' || !depsOptions.verifyInternalDeps)
+  ) {
     return;
   }
 
@@ -261,6 +266,7 @@ module.exports = {
           'peerDependencies': { 'type': ['boolean', 'array'] },
           'bundledDependencies': { 'type': ['boolean', 'array'] },
           'packageDir': { 'type': ['string', 'array'] },
+          'includeInternal': { 'type': ['boolean'] },
         },
         'additionalProperties': false,
       },
@@ -277,6 +283,7 @@ module.exports = {
       allowOptDeps: testConfig(options.optionalDependencies, filename) !== false,
       allowPeerDeps: testConfig(options.peerDependencies, filename) !== false,
       allowBundledDeps: testConfig(options.bundledDependencies, filename) !== false,
+      verifyInternalDeps: !!options.includeInternal,
     };
 
     return moduleVisitor((source, node) => {

--- a/tests/src/rules/no-extraneous-dependencies.js
+++ b/tests/src/rules/no-extraneous-dependencies.js
@@ -392,6 +392,18 @@ ruleTester.run('no-extraneous-dependencies', rule, {
         message: `'esm-package-not-in-pkg-json' should be listed in the project's dependencies. Run 'npm i -S esm-package-not-in-pkg-json' to add it`,
       }],
     }),
+
+    test({
+      code: 'import "not-a-dependency"',
+      settings: {
+        'import/resolver': { node: { paths: [ path.join(__dirname, '../../files') ] } },
+        'import/internal-regex': '^not-a-dependency.*',
+      },
+      options: [{ includeInternal: true }],
+      errors: [{
+        message: '\'not-a-dependency\' should be listed in the project\'s dependencies. Run \'npm i -S not-a-dependency\' to add it',
+      }],
+    }),
   ],
 });
 

